sympy/solvers/solveset.py

```diff
@@ -534,64 +534,104 @@ def _solve_as_rational(f, symbol, domain):
         try:
             sol = _solve_trig2(f, symbol, domain)
         except ValueError:
-            sol = sol1
-        if isinstance(sol1, ConditionSet) and isinstance(sol, ConditionSet):
-            if sol1.count_ops() < sol.count_ops():
-                sol = sol1
-    else:
-        sol = sol1
-    if sol is None:
-        raise NotImplementedError(filldedent('''
-            Solution to this kind of trigonometric equations
-            is yet to be implemented'''))
+            raise NotImplementedError(filldedent('''
+                Solution to this kind of trigonometric equations
+                is yet to be implemented'''))
     return sol
 
 
 def _solve_trig1(f, symbol, domain):
-    """Primary helper to solve trigonometric and hyperbolic equations"""
+    """Primary helper to solve trigonometric and hyperbolic equations
+
+    Either returns solution set as ConditionSet (evaluated or unevaluated)
+    or raises NotImplementedError.
+
+    """
+    # Prepare change of variable
+    x = Dummy('x')
     if _is_function_class_equation(HyperbolicFunction, f, symbol):
-        cov = exp(symbol)
+        cov = exp(x)
         inverter = invert_real if domain.is_subset(S.Reals) else invert_complex
     else:
-        cov = exp(I*symbol)
+        cov = exp(I*x)
         inverter = invert_complex
+
     f = trigsimp(f)
     f_original = f
+    trig_functions = f.atoms(TrigonometricFunction, HyperbolicFunction)
+    trig_arguments = [e.args[0] for e in trig_functions]
+    # trigsimp may have reduced the equation to an expression
+    # that is independent of 'symbol' (e.g. cos**2+sin**2)
+    if not any(a.has(symbol) for a in trig_arguments):
+        return solveset(f_original, symbol, domain)
+
+    denominators = []
+    numerators = []
+    for ar in trig_arguments:
+        try:
+            poly_ar = Poly(ar, symbol)
+        except ValueError:
+            raise ValueError("give up, we can't solve if this is not a polynomial in x")
+        if poly_ar.degree() > 1:  # degree >1 still bad
+            raise ValueError("degree of variable inside polynomial should not exceed one")
+        if poly_ar.degree() == 0:  # degree 0, don't care
+            continue
+        c = poly_ar.all_coeffs()[0]   # got the coefficient of 'symbol'
+        numerators.append(fraction(c)[0])
+        denominators.append(fraction(c)[1])
+
+    # lcm() and gcd() require more than one argument
+    if len(numerators) > 1:
+        mu = lcm(*denominators)/gcd(*numerators)
+    else:
+        mu = denominators[0]/numerators[0]
+
+    f = f.subs(symbol, mu*x)
     f = f.rewrite(exp)
     f = together(f)
     g, h = fraction(f)
     y = Dummy('y')
     g, h = g.expand(), h.expand()
     g, h = g.subs(cov, y), h.subs(cov, y)
-    if g.has(symbol) or h.has(symbol):
-        return ConditionSet(symbol, Eq(f, 0), domain)
+    if g.has(x) or h.has(x):
+        raise NotImplementedError("change of variable not possible")
```

Comment:
returning a `ConditionSet` would be more in line with the rest of `solveset`. But that entails further modifications in the upper layers: e.g. add an `is_unsolved_conditionset()` function to distinguish between `ConditionSet`s as failure condition and `ConditionSet`s as perfectly complete solution sets including validity conditions.

Reply:
Another possibility is returning a special value. If the intention is always for the exception to be caught in a specific place then I think that a specific local exception class is better as there are many places around the code base that catch or raise `NotImplementedError`:
```console
$ git grep 'raise NotImplementedError' | wc -l
     728
$ git grep 'except NotImplementedError' | wc -l
     127
```
That makes it easy for an exception to end up being unintentionally caught in the wrong place.

Reply:
Thanks for the advice. Maybe I'll just add `is_unsolved_conditionset` after all. I guess there are other places in `solveset` that are not 100% prepared to have `ConditionSet`s returned to them that _don't_ mean "I couldn't solve this". So something like that will be needed anyway and passing along Sets/ConditionSets is the established pattern in `solveset.py`.